rules/modernize-your-app/rule.md

```diff
@@ -0,0 +1,126 @@
+
+
+Pros & Cons:
+
+
+- 😃 Easy to plan
+- 😃 Pizza
+- 😡 It is still a big task, and you have re-test re-test the application to ensure that the changes have not broken anything.
+- 😡 There is no quick rollback strategy if something goes wrong.
+- 😡 BAU work must stop whilst this is happening.
+- 😡 This is not a realistic approach for most enterprise applications.
+
+
+::: ok  
+![Figure: OK example - big bang migration](big-bang.png)
+:::
+
+
+## Evolutionary
+
+
+The idea is to take the existing application and incrementally update it to a more modern architecture. This is a better approach if you have a large application that is complex and you have the time and budget to do this.
+
+
+Pros & Cons:
+
+
+- 😃 BAU development can continue on the old application.
+- 😃 You can choose the speed of the evolution - you can do it quickly or slowly, reinspecting the application at each step.
+- 😡 Can feel like yak shaving - you can end up spending a lot of time on the migration and not actually modernizing the application.
+- 😡 It is still a big task, and you generally have re-test large parts of the application to ensure that the changes have not broken anything.
+- 😡 There is no quick rollback strategy if something goes wrong.
+- 😡 At some point you'll hit a point where there needs to be a Big Bang change to get it over the line - this is not a realistic approach for most applications.
+
+
+::: ok  
+![Figure: OK example - evolutionary migration (fitting a square peg in a round hole)](evolutionary.png)
+:::
+
+
+## Strangler fig pattern (recommended)
+
+
+[Strangler Fig pattern was first described by Martin Fowler in 2004](https://martinfowler.com/bliki/StranglerFigApplication.html). The strangler fig is a type of tree that grows around other trees and slowly kills them by strangling them. This is exactly what this pattern does.
+
+
+::: no-border
+![Figure: an actual strangler fig strangling a tree](strangler-fig.jpg)
+:::
+
+
+The idea is to create a "new" application (with a modern architecture) that acts as a facade to the existing application - then port features bit by bit to the new/modern architecture. Once slices of functionality have been ported and are ready - re-point the facade to execute the new code. You can trigger this through feature flags and this also allows you to rollback to the old code if something goes wrong.
+
+
+::: tip
+Looking to incrementally update an ASP.NET application? Read about using [YARP](https://microsoft.github.io/reverse-proxy/) and [Incremental ASP.NET to ASP.NET Core Migration](https://devblogs.microsoft.com/dotnet/incremental-asp-net-to-asp-net-core-migration/)

```

Comment:
There is a tool MS made that can assist with all of this (I've also made a rule) [https://www.ssw.com.au/rules/dotnet-upgrade-assistant](https://www.ssw.com.au/rules/dotnet-upgrade-assistant). Feel free to give feedback on my rule too.

Reply:
added as a related rule